src/neural-network.js

```diff
@@ -240,47 +255,101 @@ export default class NeuralNetwork {
     let callback = options.callback;
     let callbackPeriod = options.callbackPeriod;
-    if (!options.reinforce) {
-      let sizes = [];
-      let inputSize = data[0].input.length;
-      let outputSize = data[0].output.length;
-      let hiddenSizes = this.hiddenSizes;
-      if (!hiddenSizes) {
-        sizes.push(Math.max(3, Math.floor(inputSize / 2)));
-      } else {
-        hiddenSizes.forEach(size => {
-          sizes.push(size);
-        });
-      }
-
-      sizes.unshift(inputSize);
-      sizes.push(outputSize);
+    let endTime = Date.now() + options.trainTimeMs;
+    var res = {
+      error: 1,
+      iterations: 0
+    };
 
-      this.initialize(sizes);
+    if (this.sizes === null) {
+      this.initialize(data);
     }
 
-    let error = 1;
-    let i;
-    for (i = 0; i < iterations && error > errorThresh; i++) {
+    while (res.iterations < iterations && res.error > errorThresh && Date.now () > endTime) {
+      res.iterations++;
       let sum = 0;
-      for (let j = 0; j < data.length; j++) {
-        let err = this.trainPattern(data[j].input, data[j].output, learningRate);
-        sum += err;
+      for (var i = 0; i < data.length; ++i) {
+        sum += this.trainPattern (data[i].input, data[i].output, learningRate);
       }
-      error = sum / data.length;
 
-      if (log && (i % logPeriod === 0)) {
-        log('iterations:', i, 'training error:', error);
+      res.error = sum / data.length;
+
+      if (log && (res.iterations % logPeriod === 0)) {
+        log('iterations:', res.iterations, 'training error:', res.error);
       }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
+        // JSON.parse/stringify to clone the object so the callback doesn't have side effects to training
+        callback (JSON.parse (JSON.stringify (res)));
       }
     }
+    return res;
+  }
 
-    return {
-      error: error,
-      iterations: i
+  /**
+   *
+   * @param data
+   * @param _options
+   * @param cb
+   * @returns {{error: number, iterations: number}}
+   */
+  trainAsync (data, _options = {}, cb = function() {}) {
+    if (typeof _options === "function") {
+      cb = _options;
+      _options = {};
+    }
+    const options = Object.assign({}, this.constructor.trainDefaults, _options);
+    data = this.formatData(data);
+    let iterations = options.iterations;
+    let errorThresh = options.errorThresh;
+    let log = options.log === true ? console.log : options.log;
+    let logPeriod = options.logPeriod;
+    let learningRate = _options.learningRate || this.learningRate || options.learningRate;
+    let callback = options.callback;
+    let callbackPeriod = options.callbackPeriod;
+    let endTime = Date.now() + options.trainTimeMs;
+    let res = {
+      error: 1,
+      iterations: 0
     };
+
+    if (this.sizes === null) {
+      this.initialize (data);
```

Comment:
The space after a method I've not seen in javascript libraries, can we aim for `justThis()`